engine-types/src/types/balance.rs

```diff
@@ -0,0 +1,116 @@
+use crate::fmt::Formatter;
+use crate::{Add, Display, Sub, SubAssign};
+use borsh::{BorshDeserialize, BorshSerialize};
+
+pub const ZERO_BALANCE: Balance = Balance::new(0);
+
+#[derive(
+    Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, BorshSerialize, BorshDeserialize,
+)]
+/// Engine `balance` type which wraps an underlying u128.
+pub struct Balance(u128);
+
+impl Display for Balance {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Balance {
+    /// Constructs a new `Fee` with a given u128 value.
+    pub const fn new(amount: u128) -> Balance {
+        Self(amount)
+    }
+
+    pub fn checked_sub(self, rhs: Self) -> Option<Self> {
+        self.0.checked_sub(rhs.0).map(Self)
+    }
+
+    pub fn checked_add(self, rhs: Self) -> Option<Self> {
+        self.0.checked_add(rhs.0).map(Self)
+    }
+
+    /// Consumes `Fee` and returns the underlying type.
+    pub fn into_u128(self) -> u128 {
+        self.0
+    }
```

Comment:
This should be `as_u128` as the inner value supports `Copy`. Taking self as a reference.